internal/consensus/reactor.go

```diff
@@ -292,6 +293,95 @@ conR:
 	}
 }
 
+// UpdateMetrics the applied block from the block pool updates the state metrics during the fast-sync.
+// Some implementation details please check State.recordMetrics.
+func (r *Reactor) UpdateMetrics(s sm.State, b *types.Block) {
+	rm := r.Metrics
+
+	rm.Validators.Set(float64(s.Validators.Size()))
+	rm.ValidatorsPower.Set(float64(s.Validators.TotalVotingPower()))
+
+	var (
+		missingValidators      int
+		missingValidatorsPower int64
+	)
+
+	if b.Height > s.InitialHeight {
+		var (
+			commitSize = b.LastCommit.Size()
+			valSetLen  = len(s.LastValidators.Validators)
+			address    types.Address
+		)
+		if commitSize != valSetLen {
+			panic(fmt.Sprintf("commit size (%d) doesn't match valset length (%d) at height %d\n\n%v\n\n%v",
```

Comment:
is there ever a case where, even misconfigured, monitoring/metrics code should cause the entire service to panic? 

I get that this shouldn't happen, but I think we should just return early rather than panic.

Reply:
sure, how about this place? Also just return it and give an error log?
https://github.com/tendermint/tendermint/blob/25bb556feed6e9bfdf746242e8f335dfebc6b507/internal/consensus/state.go#L1716